File: Android/app/src/main/java/com/dreamproject/inwords/viewScenario/translation/TranslationViewModelFactory.java
package com.dreamproject.inwords.viewScenario.translation;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.dreamproject.inwords.data.interactor.translation.TranslationSyncInteractor;
import com.dreamproject.inwords.data.interactor.translation.TranslationWordsInteractor;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class TranslationViewModelFactory implements ViewModelProvider.Factory {
    private final TranslationWordsInteractor translationWordsInteractor;
    private final TranslationSyncInteractor translationSyncInteractor;

    @Inject
    TranslationViewModelFactory(TranslationWordsInteractor translationWordsInteractor,
                                TranslationSyncInteractor translationSyncInteractor) {
        this.translationWordsInteractor = translationWordsInteractor;
        this.translationSyncInteractor = translationSyncInteractor;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(TranslationViewModel.class)) {
            //noinspection unchecked
            return (T) new TranslationViewModel(translationWordsInteractor, translationSyncInteractor);
        }
        throw new IllegalArgumentException("Unknown ViewModel class");
    }

}

File: Android/app/src/main/java/ru/inwords/inwords/presentation/viewScenario/translation/TranslationViewModelFactory.kt
package ru.inwords.inwords.presentation.viewScenario.translation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import ru.inwords.inwords.data.repository.texttospeech.TtsRepository

import ru.inwords.inwords.domain.interactor.translation.TranslationSyncInteractor
import ru.inwords.inwords.domain.interactor.translation.TranslationWordsInteractor
import ru.inwords.inwords.presentation.viewScenario.translation.addEditWord.AddEditWordViewModel
import ru.inwords.inwords.presentation.viewScenario.translation.translationMain.TranslationMainViewModel

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TranslationViewModelFactory @Inject
internal constructor(private val translationWordsInteractor: TranslationWordsInteractor,
                     private val translationSyncInteractor: TranslationSyncInteractor,
                     private val ttsRepository: TtsRepository) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        @Suppress("UNCHECKED_CAST")
        return when {
            modelClass.isAssignableFrom(TranslationMainViewModel::class.java) -> TranslationMainViewModel(translationWordsInteractor, translationSyncInteractor, ttsRepository)
            modelClass.isAssignableFrom(AddEditWordViewModel::class.java) -> AddEditWordViewModel(translationWordsInteractor, translationSyncInteractor)
            else -> throw IllegalArgumentException("Unknown ViewModel class")
        } as T
    }
}
